### PR TITLE
Add indexing to record_field_pat

### LINKED ISSUE #1870: Rust analyzer fails to understand tuple destructured using index
I've found a weird pattern that was used in our code base where an enum gets `match`ed using the "index notation" (I don't know how to call this...)

RLS understands that syntax but ra fails. I've created a minimal example to showcase this:

```rust
enum E {
    A(u8),
    B,
    C(u16),
}

fn check(e: E) {
    match e {
        // rust-analyzer fails here
        E::A { 0: 1 } => println!("E::A(1)"),
        // rust-analyzer fails here
        E::A { 0: 2 } => println!("E::A(2)"),
        E::A(3) => println!("E::A(3)"),
        E::A(i) => println!("E::A(i={})", i),
        E::B => println!("B"),
        E::C(4) => println!("E::C(4"),
        // rust-analyzer fails here
        E::C { 0: j } => println!("E::C(j={})", j),
    }
}

fn main() {
    check(E::A(1));
    check(E::A(2));
    check(E::A(3));
    check(E::A(4));
    check(E::B);
    check(E::C(4));
    check(E::C(5));
}
```

Screen Shot 2019-09-18 at 12 32 42 PM

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b627cf4d95693b560735cbfea4f461fd

I am using rust-analyzer d505ee968b2a99eed65dfe7be27940ad9b2647c1.

### PR DESCRIPTION
**Pull request by kjeremy**
Fixes #1870

### CODE REVIEW

File: crates/ra_parser/src/grammar/patterns.rs [hunk 1]
```diff
@@ -168,6 +168,7 @@ fn record_field_pat_list(p: &mut Parser) {
             T![.] if p.at(T![..]) => p.bump(T![..]),
 
             IDENT if p.nth(1) == T![:] => record_field_pat(p),
+            INT_NUMBER if p.nth(1) == T![:] => record_field_pat(p),
```
Comment (kjeremy) [inline crates/ra_parser/src/grammar/patterns.rs]: I thought that I could use (IDENT  || INT_NUMBER) here but that didn't work.

File: crates/ra_parser/src/grammar/patterns.rs [hunk 2]
```diff
@@ -184,12 +185,22 @@ fn record_field_pat_list(p: &mut Parser) {
...
+    assert!(p.at(IDENT) || p.at(INT_NUMBER));
     assert!(p.nth(1) == T![:]);
 
     let m = p.start();
-    name(p);
+
+    match p.current() {
+        IDENT => name(p),
+        _ => p.bump_any(),
+    }
```
Comment (matklad) [inline crates/ra_parser/src/grammar/patterns.rs]: I suggest `if !p.eat(INT_NUMBER) { name(p) }`

Comment (kjeremy) [reply crates/ra_parser/src/grammar/patterns.rs]: Done

File: crates/ra_parser/src/grammar/patterns.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_parser/src/grammar/patterns.rs]: Yeah, rust support for patterns is annoyingly non-orthogonal....

Comment (bjorn3) [reply crates/ra_parser/src/grammar/patterns.rs]: Shouldn't `IDENT | INT_NUMBER if p.nth(1) == T![:] => record_field_pat(p)` work?

Comment (kjeremy) [reply crates/ra_parser/src/grammar/patterns.rs]: Huh... so it does. I could have sworn I tried the single `|` before. I'll send up a followup PR.

Comment (bjorn3) [reply crates/ra_parser/src/grammar/patterns.rs]: > Hm, we already stabilized or patterns? Neat!

This was already supported, or patterns are like `Some(0 | 1)`, notice that the or is inside something else, not outside.

**Pull request merged**